src/lib/optimizer/strategy/between_composition_rule.cpp

```diff
@@ -0,0 +1,414 @@
+          break;
+      }
+      return {
+          left_column_expression,
+          right_column_expression,
+          type,
+          true,
+      };
+    }
+  }
+  return {
+      nullptr,
+      nullptr,
+      ColumnBoundaryType::None,
+      false,
+  };
+}
+
+const BetweenCompositionRule::ColumnBoundary BetweenCompositionRule::_create_inverse_boundary(
+    const ColumnBoundary column_boundary) const {
+  auto type = ColumnBoundaryType::None;
+  switch (column_boundary.type) {
+    case ColumnBoundaryType::UpperBoundaryInclusive:
+      type = ColumnBoundaryType::LowerBoundaryInclusive;
+      break;
+    case ColumnBoundaryType::LowerBoundaryInclusive:
+      type = ColumnBoundaryType::UpperBoundaryInclusive;
+      break;
+    case ColumnBoundaryType::UpperBoundaryExclusive:
+      type = ColumnBoundaryType::LowerBoundaryExclusive;
+      break;
+    case ColumnBoundaryType::LowerBoundaryExclusive:
+      type = ColumnBoundaryType::UpperBoundaryExclusive;
+      break;
+    default:
+      break;
+  }
+
+  return {
+      std::static_pointer_cast<LQPColumnExpression>(column_boundary.border_expression),
+      std::static_pointer_cast<AbstractExpression>(column_boundary.column_expression),
+      type,
+      true,
+  };
+}
+
+static PredicateCondition get_between_predicate_condition(bool left_inclusive, bool right_inclusive) {
+  if (left_inclusive && right_inclusive) {
+    return PredicateCondition::BetweenInclusive;
+  } else if (left_inclusive && !right_inclusive) {
+    return PredicateCondition::BetweenUpperExclusive;
+  } else if (!left_inclusive && right_inclusive) {
+    return PredicateCondition::BetweenLowerExclusive;
+  } else if (!left_inclusive && !right_inclusive) {
+    return PredicateCondition::BetweenExclusive;
+  }
+  Fail("Unreachable Case");
+}
+
+/**
+ * _replace_predicates gets a vector of AbstractLQPNodes as input and
+ * substitutes suitable BinaryPredicateExpressions with BetweenExpressions.
+ * Furthermore BinaryPredicateExpressions which are obsolete after the substitution
+ * are removed.
+**/
+void BetweenCompositionRule::_replace_predicates(std::vector<std::shared_ptr<AbstractLQPNode>>& predicates) const {
+  // Store original input and output
+  auto input = predicates.back()->left_input();
+  const auto outputs = predicates.front()->outputs();
+  const auto input_sides = predicates.front()->get_input_sides();
+
+  auto between_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>();
+  auto predicate_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>();
+  std::unordered_map<LQPColumnReference, std::vector<ColumnBoundary>> column_boundaries;
+
+  // Filter predicates with a boundary to the boundaries vector
+  for (auto& predicate : predicates) {
+    // A logical expression can contain multiple binary predicate expressions
+    std::vector<std::shared_ptr<BinaryPredicateExpression>> expressions;
+    const auto predicate_node = std::static_pointer_cast<PredicateNode>(predicate);
```

Comment:
The data type of the provided `predicates` is `AbstractLQPNode` which is also stated by the method comment. However, the calling method ensures that all provided nodes are `PredicateNodes`, right? Should we clarify that we can safely use a static cast here or should we do the cast in the calling method and provide a vector or `PredicateNodes`?